### PR TITLE
[mongodb]Connect to <allocated ip>:27017

### PR DESCRIPTION
**Pull request by rawler**
Should either connect to localhost:_allocated port_ or _allocated ip_:27017.

_allocated ip_:_random port_ is most likely not listening.

### CODE REVIEW

**Comment by rawler** [comment]
As a side-note, what is the rationale behind binding to host-ports at all?

File: mongodb.go [hunk 1]
```diff
@@ -7,33 +7,34 @@ import (
...
-	c, ip, port, err := SetupMongoContainer()
+	c, ip, _, err := SetupMongoContainer()
 	if err != nil {
 		return c, fmt.Errorf("Could not set up MongoDB container: %v", err)
 	}
 
 	for try := 0; try <= tries; try++ {
 		time.Sleep(delay)
-		url := fmt.Sprintf("%s:%d", ip, port)
+		url := fmt.Sprintf("%s:%d", ip, mongoDBPort)
```
Comment (arekkas) [inline mongodb.go]: Why use a constant here instead of reyling on the return value?

File: mongodb.go [hunk 2]
```diff
@@ -7,33 +7,34 @@ import (
...
 	port = RandomPort()
-	forward := fmt.Sprintf("%d:%d", port, 27017)
+	forward := fmt.Sprintf("%d:%d", port, mongoDBPort)
 	if BindDockerToLocalhost != "" {
 		forward = "127.0.0.1:" + forward
 	}
-	c, ip, err = SetupContainer(MongoDBImageName, port, 10*time.Second, func() (string, error) {
-		res, err := run("--name", GenerateContainerID(), "-d", "-P", "-p", forward, MongoDBImageName)
-		return res, err
+	c, ip, err = SetupContainer(MongoDBImageName, mongoDBPort, 10*time.Second, func() (string, error) {
```
Comment (arekkas) [inline mongodb.go]: I'm not sure that replacing the random "outside" port with the static "inside" port will work. What was your intention?

**Comment by arekkas** [comment]
> As a side-note, what is the rationale behind binding to host-ports at all?

Not sure what you mean by this. Could you elaborate?

Regarding the code changes: I'm not sure what you're trying to achieve but I think that the PR will break things. Could you maybe explain what your problem is?

**Comment by arekkas** [comment]
also could you please rebase against master? that will fix travis ci

**Comment by rawler** [comment]
OK, so I now understand the rationale for exposing to public ports. I guess the use for dockertest is geared towards running tests outside the docker-host, I.E. on a docker-swarm or similar?

I would like to run my test _on_ the docker-host, I.E. using docker-in-docker, or on a VM dedicated to Q/A being it's own docker-host.

In that case, I have the host-ip, and an allocated docker-ip for the MongoDB instance. MongoDB is available on _host-ip_:_allocated-port_ and on _docker-ip_:_27017_. However, SetupMongoContainer returns _docker-ip_ and _allocated-port_, which on line 36 is concatenated, yielding an URL where MongoDB is definitely _not_ listening.

Likewise, on line 19, SetupMongoContainer is called with the RandomPort()-argument, but c.lookup() called in docker.go:230 seems to want the port of the service on the allocated container-ip. 27017, that is.

I thought it was merely a typo in the MongoDB-code, but I guess there is some more general problem with how I'm running docker?

The actual Error I'm getting is "Could not set up MongoDB container: 192.168.128.2:6203 unreachable for 10s", where 192.168.128.2 is the IP of the container, not the host.

**Comment by rawler** [comment]
Seems to work with BindDockerToLocalhost set. I guess this is closely related to the decision in container.go:51. In the last branch, the actual service IP should be used, not a bound public port. (In fact the entire port binding is unecessary in that step). I will close this PR and see if I can whip up something more general to better handle this.

**Comment by arekkas** [comment]
The issue is that docker-machine needs explicit port forwarding when you want to access through `localhost`. If docker is ran from within linux, you don't need that. We could probably just add a check for that and save the pain of seeting that up yourself.

**Pull request closed without merging**